M2/Macaulay2/m2/res.m2

```diff
@@ -154,103 +203,179 @@ resolutionInEngine := opts -> (M) -> (
+	DegreeLimit		=> null,	-- slant degree limit
+	SyzygyLimit		=> infinity,	-- number of min syzs found
+	PairLimit		=> infinity,	-- number of pairs computed
+	HardDegreeLimit		=> {},		-- throw out information in degrees above this one
+	-- HardLengthLimit	=> infinity,	-- throw out information in lengths above this one
+	SortStrategy		=> 0,		-- strategy choice for sorting S-pairs
+	Strategy		=> null,	-- algorithm to use, usually 1, but sometimes 2
+	FastNonminimal		=> false
+	}
+    )
 
-engineReady := M -> (
-     R := ring M;
-     -- Needed to compute resolutions, (algorithms 0,1,2,3):
-     --    Ring is (tower of) poly ring(s) over a field (or skew commutative, or quotient ring of such, or both)
-     --    Ring is graded
-     --    Ring is homogeneous in this grading
-     --    Matrix is homogeneous in this grading
-     -- Additional requirements for resolution algorithm 3 (which uses hilbert function):
-     --    Ring is singly graded
-     R.?Engine 
-     and isHomogeneous M
-     and (isCommutative R or isSkewCommutative R)
-     and (
-     	  k := ultimate(coefficientRing, R);
-	  k =!= R
-     	  and isField k
-	  )
-     )
-
-protect ManualResolution				    -- not to be exported
+-- keys: none so far
+-- TODO: perhaps keys for different types of resolutions? e.g. virtual resolution?
+ResolutionContext = new SelfInitializingType of Context
+ResolutionContext.synonym = "resolution context"
+
+new ResolutionContext from Module := (C, M) -> new C from {}
+
+-- keys: LengthLimit
+-- TODO: what else?
+-- SyzygyLimit, HardDegreeLimit, StopBeforeComputation,
+-- DegreeLimit, FastNonminimal, SortStrategy, PairLimit
+ResolutionComputation = new Type of Computation
+ResolutionComputation.synonym = "resolution computation"
+
+new ResolutionComputation from HashTable := (C, H) -> merge(H, new HashTable from { Result => null }, last)
+
+isComputationDone Resolution := Boolean => options resolution >> opts -> W -> (
+    -- can returnCode ever be not defined?
+    isComputationDone W.returnCode
+    and opts.LengthLimit <= W.length
+    and opts.DegreeLimit <= W.DegreeLimit)
+
+isComputationDone ResolutionComputation := Boolean => options resolution >> opts -> container -> (
+    -- this function determines whether we can use the cached result, or further computation is necessary
+    C := if instance(container.Result, ChainComplex) then container.Result else return false;
+    -- if FastNonminimal sets returnCode correctly, the "not container.FastNonminimal" condition can be removed
+    if C.?Resolution and not container.FastNonminimal then isComputationDone(C.Resolution, opts) else
+    ( opts.FastNonminimal == container.FastNonminimal or not container.FastNonminimal )
+    and opts.DegreeLimit  <= container.DegreeLimit
+    and opts.LengthLimit  <= container.LengthLimit)
+
+updateComputation(ResolutionComputation, ChainComplex) := ChainComplex => options resolution >> opts -> (container, result) -> (
```

Comment:
I'm a bit worried about the logic in updateComputation (although with the current behavior, I think it is fine).  Suppose we compute with a low degree, but large length, and then switch it to high degree, low length.  I know that currently, we are just redoing the computation anyway, but suppose the engine can deal with increasing either one (which it will do before too long, I hope).  Then this will not reflect what is currently computed.  I guess in that case we just want to modify the values e.g. container.DegreeLimit, if e.g. opts.DegreeLimit is larger than it was.

I also think that (for the res.m2 rewrite) FastNonminimal resolutions are really different from minimal resolutions.  For instance, should we allow use of these resolutions for computing Ext?

Reply:
This is a great point, and I thought about it quite a bit. The logic of `isComputationDone` is to determine whether what is cached is enough to get the result without any further engine level computations (sometimes a little top-level computation is necessary, and `adjustComputation` helps with that). If `isComputationDone` determines that we're asking for either more degrees or more terms, then we make some engine calls, and at that point the engine can either recompute from scratch or increase either or both, so I think this case is handled well by the existing code (e.g. `M.cache.resolution.Resolution` stuff), not even mine.

Hopefully when this is brought into `Complexes` we can simplify this further though.

> I also think that (for the res.m2 rewrite) FastNonminimal resolutions are really different from minimal resolutions. For instance, should we allow use of these resolutions for computing Ext?

I think currently minimal resolutions replace cached `FastNonminimal` resolutions when requested. Should I change it to keep both cached?

Reply:
I'll be rewriting res.m2 quite a bit after this pull request is accepted.  I suggest we wait on making any changes until I do that.